base/expr.jl

```diff
@@ -355,6 +355,49 @@ macro noinline(x)
     return annotate_meta_def_or_block(x, :noinline)
 end
 
+"""
+    @outline expr
+
+Outline an expression into its own function, and call that function.
+
+This macro introduces a "function barrier", which can be helpful in some code optimization
+scenarios. The expr is extracted into an outlined function, which is marked `@noinline`.
+
+Outlining an expr can be used to make a function smaller, e.g. by outlining an unlikely
+branch, which could help with runtime performance by improving instruction cache locality,
+and could help with compilation performance since two smaller functions can sometimes compile
+faster than one larger function. Finally, outlining can be useful for type-stability, by
+outlining a type unstable block within a hot loop, where the outlined function could be type
+stable.
+
+A common use case is to `@outline` the code that throws exceptions, since this should be a
+rare case, but it can introduce a lot of complexity to the generated code, which can
+sometimes harm the compiler's ability to optimize.
+
+# Examples
+```julia
+function getindex(container, index)
+    if index < 1 || index > length(container)
+        # Outline this throw, since constructing a BoundsError requires boxing
+        # the arguments, which produces a lot of code.
+        @outline throw(BoundsError(container, index))
+    end
+    return container.data[index]
+end
+```
+"""
+macro outline(expr)
+    vars = esc.(_free_vars(expr))
+    quote
+        @noinline outline($(vars...)) = $(esc(expr))
+
+        outline($(vars...))
+    end
+end
+_free_vars(s::Symbol) = [s]
+_free_vars(_) = []
+_free_vars(e::Expr) = isempty(e.args) ? [] : unique!(mapreduce(_free_vars, vcat, e.args))
```

Comment:
I think `unique!` is not defined yet so this fails to compile for me

although, what happens if you use `outline` as a zero-arg closure and don't pass the vars in explicitly at all? like

```
macro outline2(expr)
    local fname = gensym(:outlined_expr)
    quote
        @noinline $(fname)() = $(esc(expr))
        $(fname)()
    end
end
```

comparing the `@code_native` of
```
foo1() = @outline rand() > 0.5 || throw(AssertionError("abc"))
foo2() = @outline2 rand() > 0.5 || throw(AssertionError("abc"))
```

it seems a lot simpler, but I'm not an expert at these things so maybe I'm missing something

Reply:
One difference is in how they handle type instability:
```julia-repl
julia> function bar()
           x = rand(Bool) ? 7 : 7.0
           @outline x*x+x*x+x*x*x+x/x-x+x
       end
bar (generic function with 1 method)

julia> function bar2()
           x = rand(Bool) ? 7 : 7.0
           @outline2 x*x+x*x+x*x*x+x/x-x+x
       end
bar2 (generic function with 1 method)

julia> @b bar
6.373 ns

julia> @b bar2
75.544 ns (2.40 allocs: 38.346 bytes)
```

Reply:
Yeah, the issue with 0-arg closures is that they can box their arguments in a bunch of caess. I was hoping to avoid that by taking advantage of macro-magic. :)

Reply:
Thanks for the bug report though - shouldn't be too hard to fix. I didn't notice working locally with Revise.